content/all/react-testing-4-cypress.md

```diff
@@ -15,40 +15,42 @@ tags:
+- [Modern React testing, part 2: Jest and Enzyme](/all/react-testing-2-jest-and-enzyme/)
+- [Modern React testing, part 3: Jest and React Testing Library](/all/react-testing-3-jest-and-react-testing-library/)
+- **Modern React testing, part 4: Cypress and Cypress Testing Library (_this post_)**
+
+Check out [the GitHub repository](https://github.com/sapegin/cypress-article-2020) with all the examples.
 
 ## Getting started with Cypress and Cypress Testing Library
 
 We’ll set up and use these tools:
 
 - [Cypress](https://www.cypress.io/), an end-to-end test runner;
-- [Cypress Testing Library](https://github.com/testing-library/cypress-testing-library), additional queries.
+- [Cypress Testing Library](https://github.com/testing-library/cypress-testing-library), additional semantic queries.
 - [Mock Service Worker](https://mswjs.io/), mocking network requests.
-- [start-server-and-test](https://www.npmjs.com/package/start-server-and-test), starts server, waits for URL, then runs test command.
+- [start-server-and-test](https://github.com/bahmutov/start-server-and-test), starts a server, waits for the URL, then runs the test command.
 
 ### Why Cypress and Cypress Testing Library
 
 **Cypress** has many benefits over other end-to-end test runners:
 
 - The best experience of writing and debugging tests.
-- Inspect the page at any moment during the test run using the browser developer tools.
-- All commands wait for the DOM to change when necessary, this makes writing async tests easier.
-- Assertions better resemble real user behavior, like checking that the button is visible, isn’t disabled, and isn’t hidden behind another element before clicking it.
+- Ability to inspect the page at any moment during the test run using the browser developer tools.
+- All commands wait for the DOM to change when necessary, which simplifies testing async behavior.
+- Tests better resemble real user behavior. For example, Cypress checks that a button is visible, isn’t disabled, and isn’t hidden behind another element before clicking it.
 - Supports Chrome, Firefox and Edge.
 
 **Cypress Testing Library** makes Cypress even better:
 
-- Convenient queries, like finding elements by their label text or ARIA role.
+- Convenient semantic queries, like finding elements by their label text or ARIA role.
 - Libraries for other frameworks with the same queries.
 
-Testing Library helps you write good tests and makes writing bad tests difficult. It allows you to query elements similar to how a user would do that: for example, query form elements and buttons by their labels.
+Testing Library helps us write [good tests](/all/react-testing-1-best-practices/) and makes writing bad tests difficult. It allows us to interact with the app similar to how a real user would do that: for example, find form elements and buttons by their labels. It helps us to avoid testing implementation details, making our tests resilient to code changes that don’t change the behavior.
```

Comment:
and makes it difficult to write bad tests?

Reply:
Why do you prefer a more complex variant?